### ISSUE TITLE
Mac error with imagemagick 6.7.7-6

### ISSUE THREAD

**Issue by max-mapper** [z24/tty2gif#1]
here is a full output:

``` sh
$ make
g++ -O -I/usr/local/Cellar/imagemagick/6.7.7-6/include/ImageMagick  -o tty2gif tty2gif.cpp -L/usr/local/Cellar/imagemagick/6.7.7-6/lib -lMagick++ -lMagickCore 
$ ./tty2gif foo.raw bar.gif
bash-3.2$ hi
bash: hi: command not found
bash-3.2$ exit
exit


Assertion failed: (wid && *wid), function SaveReplay, file tty2gif.cpp, line 203.
                                                                                 Abort trap: 6
$ 
```

**Comment by rafi** [comment]
Any clue? On OSX with X. Maybe should set lib/include/prefix?

**Comment by z24** [comment]
@rafi Hey, there. I have tried compiling on a Mac VM. I have installed ImageMagick using MacPorts, and installed the xquartz for X11. 

You need to replace `g++` in Makefile with `clang++`. Then I can compile it. BUT `tty2gif` somehow can not grab x window handler. I will take a look if I have time later.

**Comment by PurrBiscuit** [comment]
Hi there, similar error to the two folks who posted above, on a mac running OSX v10.9.5.  Installed Imagemagick using homebrew.

```
Stephens-MacBook-Pro-2:~/Desktop/z24-tty2gif-dad1a8f $ ./tty2gif hi.raw out.gif
~/Desktop/z24-tty2gif-dad1a8f $ ls -a
.		..		Makefile	README.md	hi.raw		tty2gif		tty2gif.cpp
~/Desktop/z24-tty2gif-dad1a8f $ echo "hello"
hello
~/Desktop/z24-tty2gif-dad1a8f $ exit
exit
Assertion failed: (wid && *wid), function SaveReplay, file tty2gif.cpp, line 203.
                                                                                 Abort trap: 6
```

I'd love to use this but haven't been able to figure out the fix for this.  Any ideas?  Thank you.

**Comment by jirutka** [comment]
The same problem here; OS X 10.9.5, ImageMagick 6.9.1-0 (installed via Homebrew), XQuartz 2.7.7.

**Comment by z24** [comment]
Hi all, sorry for the long absence. Just patch the code with a new release. Check it now to see whether it will work or not on Mac. I have tested on my VM and it works fine.

@jirutka @PurrBiscuit I installed ImageMagick via macports, and I hope `pkg-config` should find correct paths if installed via Homebrew, otherwise, you may consider manually modify the Makefile.

**Comment by jirutka** [comment]
Now it can’t find X11.

```sh
 tty2gif git:(master) make
clang++ -O2 -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -DMAGICKCORE_HDRI_ENABLE=0 -DMAGICKCORE_QUANTUM_DEPTH=16 -I/usr/local/Cellar/imagemagick/6.9.1-0/include/ImageMagick-6  -I/opt/X11/include -o tty2gif tty2gif.cpp -L/usr/local/Cellar/imagemagick/6.9.1-0/lib -lMagick++-6.Q16 -lMagickWand-6.Q16 -lMagickCore-6.Q16  -lX11
ld: library not found for -lX11
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [tty2gif] Error 1
```

**Comment by z24** [comment]
@jirutka Eh...my bad, a quick turnaround: `CFLAGS += -I/opt/X11/include -L/opt/X11/lib`

**Comment by jirutka** [comment]
It compiles fine now, but generating of gif fails:

```sh
$ ./tty2gif out.raw out.gif
$ ll
$ exit
libc++abi.dylib: terminating with uncaught exception of type Magick::ErrorBlob: tty2gif: unable to open image `0x1': No such file or directory @ error/blob.c/OpenBlob/2675
                     [1]    32079 abort      ./tty2gif out.raw out.gif
```

**Comment by z24** [comment]
@jirutka Are you testing it under a `xterm`? Otherwise, it will fail.

**Comment by jirutka** [comment]
Nope, I’m using iTerm2. I’ve tried to run it directly under xterm from XQuartz, but it failed as well.

**Comment by z24** [comment]
@jirutka That's weird. Could you post error message when you run it under `xterm`?

**Comment by jirutka** [comment]
Yeah, but it’s the same as when I run it under iTerm2.
```
libc++abi.dylib: terminating with uncaught exception of type Magick::ErrorBlob: tty2gif: unable to open image `0x60000d': No such file or directory @ error/blob.c/OpenBlob/2675
                [1]    37901 abort      ./tty2gif out.raw out.gif
```

**Comment by z24** [comment]
@jirutka Truly thanks for your help. It seems your ImageMagick do not support X11. Since I do not have homebrew, I can only guess possible issues. Do you have `display` command in your installed ImageMagick package? If not, probably your ImageMagick does not support X11. You need to re-install it, check this [page](http://paulmestemaker.com/imagemagick-display-mac-os-x/).

**Comment by jirutka** [comment]
I thank you. :) You’re right, I’ve reinstalled imagemagick with x11 and the previous error is gone. However, it still doesn’t work, now it fails with:

```
tty2gif: unable to read X window image `0x60000d': Resource temporarily unavailable @ error/xwindow.c/XImportImage/4984.
                                        libc++abi.dylib: terminating with uncaught exception of type Magick::WarningImage: tty2gif: No image was loaded.
                                                                        [1]    65914 abort      ./tty2gif out.raw out.gif
```

**Comment by z24** [comment]
@jirutka Could you try [this script](https://gist.github.com/z24/b8c6c72473e825b3e2fd)? This script will test your ImageMagick on taking screenshot on a X11 window. It will ask you to click on a target X11 window, and you need to select on any X11 window, e.x., a xterm window. After that a `out.png` should be generated, check this file whether it is the screenshot of the selected window, or report error messages if any. Thanks!

**Comment by jirutka** [comment]
@z24 Sorry for long delay. I’ve tested the script and it works, i.e. `out.png` is created.

**Comment by z24** [comment]
@jirutka That's unexpected. Because, to capture the screenshots, `tty2gif` actually follows the similar logic implemented in that script. If I have time I will test it under brew installed ImageMagick.

**Comment by MaryanMorel** [comment]
Same error here, with OS X 10.11.2, XQuartz 2.7.8, imagemagick installed with brew (using `--with-x11` flag). The script you provide works and produces out.png.

**Comment by zackse** [comment]
There seems to be a race condition with getting access to the X window, but I haven't been able to track it down. I tried switching to the MagickWand API (https://github.com/zackse/tty2gif/commit/75bd65aaeeab32bba32e94df7bac98749b5b60f1), but the same error occurs.

Adding a sleep before attempting to read the X image, however, seems to work reliably on my machine, though the fix is quite unsatisfying:

https://github.com/zackse/tty2gif/commit/0002ca708d69ea9f085a527be518fbb7b1590223

I discovered the fix when running the code through the debugger -- the delay from stopping at breakpoints prevented the error from occurring. I wasn't able to step into the ImageMagick code to figure out what's breaking in `XGetWindowImage`, but maybe this gives someone an idea for a better solution.

**Comment by zackse** [comment]
The problem is that some calls to `XGetImage` are failing, apparently when they happen too quickly. The relevant code is in [ImageMagick's XGetWindowImage function](https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/xwindow.c#L4227), which is called from `XImportImage`.

I'm not sure if this is an XQuartz problem, because I don't see this behavior on Linux (Ubuntu).

For reference, my environment is OS X 10.12.3, ImageMagick 7.0.4-6, clang-800.0.42.1.

**Comment by MaryanMorel** [comment]
Alternative on macOS: https://github.com/icholy/ttygif
